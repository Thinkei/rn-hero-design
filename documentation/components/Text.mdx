---
name: Text
menu: Components
---

import { Playground } from 'docz'
import { Text } from 'rn-hero-design'

# Text

## Basic usage

<Playground>
  <Text size="h1">The quick brown fox jumps over the lazy dog</Text>
  <br/>
  <Text size="h2">The quick brown fox jumps over the lazy dog</Text>
  <br/>
  <Text size="h3">The quick brown fox jumps over the lazy dog</Text>
  <br/>
  <Text size="h4">The quick brown fox jumps over the lazy dog</Text>
  <br/>
  <Text size="h5">The quick brown fox jumps over the lazy dog</Text>
</Playground>

## Properties

