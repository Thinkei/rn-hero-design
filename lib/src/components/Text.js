import React from 'react';
import t from 'prop-types';
import { Text as RNText, StyleSheet } from 'react-native';
import heroTheme from '../themes/hero';

const Text = ({ children, size, weight, color, style, theme = heroTheme }) => (
  <RNText
    style={StyleSheet.flatten([
      theme.text.text,
      theme.text[size],
      { fontWeight: weight, color },
      style,
    ])}
  >
    {children}
  </RNText>
);

Text.propTypes = {
  children: t.string,
  size: t.oneOf(['h1', 'h2', 'h3', 'h4', 'h5']),
  weight: t.oneOf(['300', '400', '500', '600', '700']),
  color: t.string,
  style: t.object,
  /**
   * The theme object contains style of all components. Can be injected by `injectTheme` HOC
   */
  theme: t.object,
};

export default Text;
