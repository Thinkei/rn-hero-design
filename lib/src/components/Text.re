module RN = ReactNative;
module RNText = RN.Text;

let noStyle = RN.Style.style();

[@bs.deriving jsConverter]
type size = [ | `h1 | `h2 | `h3 | `h4 | `h5];

[@bs.deriving jsConverter]
type weight = [
  | [@bs.as "300"] `_300
  | [@bs.as "400"] `_400
  | [@bs.as "500"] `_500
  | [@bs.as "600"] `_600
  | [@bs.as "700"] `_700
];

let toRNWeight =
  fun
  | `_300 => `_300
  | `_400 => `_400
  | `_500 => `_500
  | `_600 => `_600
  | `_700 => `_700;

let (|?) = (x, y) =>
  switch (x) {
  | None => y
  | Some(x) => x
  };

[@react.component]
let make =
    (
      ~children,
      ~style=?,
      ~size=?,
      ~weight=?,
      ~color=?,
      ~theme=Hero_Theme.default,
    ) =>
  <RNText
    style={RN.StyleSheet.flatten([|
      theme##text##styles##text,
      Belt.Option.mapWithDefault(size, noStyle, size =>
        switch (size) {
        | `h1 => theme##text##styles##h1
        | `h2 => theme##text##styles##h2
        | `h3 => theme##text##styles##h3
        | `h4 => theme##text##styles##h4
        | `h5 => theme##text##styles##h5
        }
      ),
      RN.Style.style(
        ~fontWeight=?Belt.Option.map(weight, toRNWeight),
        ~color?,
        (),
      ),
      style |? noStyle,
    |])}>
    {children |> React.string}
  </RNText>;

let default = make;
