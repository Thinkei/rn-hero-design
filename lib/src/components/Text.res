open ReactNative

let emptyStyle = Style.style()

@genType @deriving(jsConverter)
type weight = [
  | @as("100") #_100
  | @as("200") #_200
  | @as("300") #_300
  | @as("400") #_400
  | @as("500") #_500
  | @as("600") #_600
  | @as("700") #_700
  | @as("800") #_800
  | @as("900") #_900
  | #bold
  | #normal
]

@genType @deriving(jsConverter)
type ellipsizeMode = [#clip | #head | #middle | #tail]

let \"|?" = (x, y) =>
  switch x {
  | Some(x) => x
  | None => y
  }

@genType @deriving(jsConverter)
type size = [#h1 | #h2 | #h3 | #h4 | #h5]

@genType @react.component
let make = (
  ~testID=?,
  ~children,
  ~size: option<size>=?,
  ~weight=?,
  ~color=?,
  ~ellipsizeMode=?,
  ~numberOfLines=?,
  ~onPress=?,
  ~style=?,
  ~theme=Hero_Theme.default,
) =>
  <ReactNative.Text
    ?testID
    ellipsizeMode=?{ellipsizeMode->Belt.Option.flatMap(ellipsizeModeFromJs)}
    ?numberOfLines
    ?onPress
    style={StyleSheet.flatten([
      theme["text"]["text"],
      Belt.Option.mapWithDefault(size, emptyStyle, size =>
        switch size {
        | #h1 => theme["text"]["h1"]
        | #h2 => theme["text"]["h2"]
        | #h3 => theme["text"]["h3"]
        | #h4 => theme["text"]["h4"]
        | #h5 => theme["text"]["h5"]
        | _ => emptyStyle
        }
      ),
      Style.style(
        ~fontWeight=weight->Belt.Option.flatMap(weightFromJs)->Belt.Option.getWithDefault(#_400),
        ~color?,
        (),
      ),
      \"|?"(style, emptyStyle),
    ])}>
    {children |> React.string}
  </ReactNative.Text>

@genType
let default = Helpers.injectTheme(make)
