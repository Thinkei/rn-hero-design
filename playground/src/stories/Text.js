import React from 'react';
import { ScrollView } from 'react-native';
import { Text, Container, injectTheme } from 'rn-hero-design';

const TextScreen = ({ theme }) => (
  <ScrollView>
    <Container>
      <Text size="h1" style={{ marginBottom: theme.variables.MEDIUM_SIZE }}>
        Text size h1
      </Text>

      <Text size="h2" style={{ marginBottom: theme.variables.MEDIUM_SIZE }}>
        Text size h2
      </Text>

      <Text size="h3" style={{ marginBottom: theme.variables.MEDIUM_SIZE }}>
        Text size h3
      </Text>

      <Text size="h4" style={{ marginBottom: theme.variables.MEDIUM_SIZE }}>
        Text size h4
      </Text>

      <Text size="h5" style={{ marginBottom: theme.variables.MEDIUM_SIZE }}>
        Text size h5
      </Text>

      <Text weight="500" style={{ marginBottom: theme.variables.MEDIUM_SIZE }}>
        Text weight 500
      </Text>

      <Text weight="600" style={{ marginBottom: theme.variables.MEDIUM_SIZE }}>
        Text weight 600
      </Text>

      <Text weight="700" style={{ marginBottom: theme.variables.MEDIUM_SIZE }}>
        Text weight 700
      </Text>

      <Text
        numberOfLines={1}
        ellipsizeMode="middle"
        style={{ marginBottom: theme.variables.MEDIUM_SIZE }}>
        Long paragraph, supports numberOfLines and ellipsizeMode
      </Text>
    </Container>
  </ScrollView>
);

export default injectTheme(TextScreen);
