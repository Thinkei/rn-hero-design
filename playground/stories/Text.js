import React from 'react';
import { ScrollView } from 'react-native';
import { Text, ReText } from 'rn-hero-design';

const TextScreen = () => (
  <ScrollView>
    <Text size="h1" style={{ marginBottom: 16 }}>
      🐔 The quick brown fox
    </Text>
    <Text size="h2" style={{ marginBottom: 16 }}>
      🐔 The quick brown fox
    </Text>
    <Text size="h3" style={{ marginBottom: 16 }}>
      🐔 The quick brown fox
    </Text>
    <Text style={{ marginBottom: 16 }}>🐔 The quick brown fox</Text>
    <Text weight="500" style={{ marginBottom: 16 }}>
      🐔 The quick brown fox
    </Text>
    <Text weight="600" style={{ marginBottom: 16 }}>
      🐔 The quick brown fox
    </Text>
    <Text weight="700" style={{ marginBottom: 32 }}>
      🐔 The quick brown fox
    </Text>

    <ReText size={'h1'} style={{ marginBottom: 16 }}>
      🐪 The quick brown fox
    </ReText>
    <ReText size={'h2'} style={{ marginBottom: 16 }}>
      🐪 The quick brown fox
    </ReText>
    <ReText size={'h3'} style={{ marginBottom: 16 }}>
      🐪 The quick brown fox
    </ReText>
    <ReText style={{ marginBottom: 16 }}>🐪 The quick brown fox</ReText>
    <ReText weight={'500'} style={{ marginBottom: 16 }}>
      🐪 The quick brown fox
    </ReText>
    <ReText weight={'600'} style={{ marginBottom: 16 }}>
      🐪 The quick brown fox
    </ReText>
    <ReText weight={'700'} style={{ marginBottom: 16 }}>
      🐪 The quick brown fox
    </ReText>
  </ScrollView>
);

export default TextScreen;
