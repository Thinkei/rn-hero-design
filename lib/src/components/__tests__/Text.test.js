import React from 'react';

import { render } from '@testing-library/react-native';

import Text from '../Text.bs';

describe('Text', () => {
  it.each`
    element
    ${<Text size="h1" weight="100" ellipsizeMode="clip" numberOfLines={2} onPress={() => {}} />}
    ${<Text size="h2" weight="normal" ellipsizeMode="clip" numberOfLines={2} onPress={() => {}} />}
    ${<Text size="h3" weight="bold" ellipsizeMode="clip" numberOfLines={2} onPress={() => {}} />}
    ${<Text size="h4" weight="600" ellipsizeMode="clip" numberOfLines={2} onPress={() => {}} />}
    ${<Text size="h5" weight="1000" ellipsizeMode="clip" numberOfLines={2} onPress={() => {}} />}
  `('renders correctly', ({ element }) => {
    const { baseElement } = render(element);
    expect(baseElement).toMatchSnapshot();
  });
});
