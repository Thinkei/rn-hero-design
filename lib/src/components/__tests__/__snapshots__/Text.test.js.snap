// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text renders correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Text
      ellipsizeMode="clip"
      numberOfLines={2}
      style={
        Object {
          "color": "#000000",
          "fontFamily": "Proxima Nova",
          "fontSize": 32.5,
          "fontWeight": "100",
        }
      }
    />
  </View>
</View>
`;

exports[`Text renders correctly 2`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Text
      ellipsizeMode="clip"
      numberOfLines={2}
      style={
        Object {
          "color": "#000000",
          "fontFamily": "Proxima Nova",
          "fontSize": 24.375,
          "fontWeight": "normal",
        }
      }
    />
  </View>
</View>
`;

exports[`Text renders correctly 3`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Text
      ellipsizeMode="clip"
      numberOfLines={2}
      style={
        Object {
          "color": "#000000",
          "fontFamily": "Proxima Nova",
          "fontSize": 19.0125,
          "fontWeight": "bold",
        }
      }
    />
  </View>
</View>
`;

exports[`Text renders correctly 4`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Text
      ellipsizeMode="clip"
      numberOfLines={2}
      style={
        Object {
          "color": "#000000",
          "fontFamily": "Proxima Nova",
          "fontSize": 16.25,
          "fontWeight": "600",
        }
      }
    />
  </View>
</View>
`;

exports[`Text renders correctly 5`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Text
      ellipsizeMode="clip"
      numberOfLines={2}
      style={
        Object {
          "color": "#000000",
          "fontFamily": "Proxima Nova",
          "fontSize": 13.487499999999999,
          "fontWeight": "400",
        }
      }
    />
  </View>
</View>
`;
